src/CoreFoundation/CFProxySupport.cs

```diff
@@ -580,61 +580,105 @@ public static CFProxySettings GetSystemProxySettings ()
+						var array = new CFArray (proxyList, false);
+						proxies = new CFProxy [array.Count];
+						for (int i = 0; i < proxies.Length; i++) {
+							var dict = new NSDictionary (array.GetValue (i));
+							proxies[i] = new CFProxy (dict);
+						}
+						array.Dispose ();
+					}
+					runLoop.Stop ();
+				};
+
+				var clientContext = new CFStreamClientContext ();
+				using (var loopSource = new CFRunLoopSource (CFNetworkExecuteProxyAutoConfigurationScript (pacScript.Handle, url.Handle, cb, ref clientContext)))
+				using (var mode = new NSString ("Xamarin.iOS.Proxy")) {
+					runLoop.AddSource (loopSource, mode);
+					runLoop.RunInMode (mode, double.MaxValue, false);
+					runLoop.RemoveSource (loopSource, mode);
+					return proxies;
+				}
+			}
 		}
 		
 		[DllImport (Constants.CFNetworkLibrary)]
 		extern static /* CFRunLoopSourceRef __nonnull */ IntPtr CFNetworkExecuteProxyAutoConfigurationURL (
 			/* CFURLRef __nonnull */ IntPtr proxyAutoConfigurationURL,
 			/* CFURLRef __nonnull */ IntPtr targetURL,
-			/* CFProxyAutoConfigurationResultCallback __nonnull */ IntPtr cb,
-			/* CFStreamClientContext * __nonnull */ IntPtr clientContext);
+			/* CFProxyAutoConfigurationResultCallback __nonnull */ CFProxyAutoConfigurationResultCallback cb,
+			/* CFStreamClientContext * __nonnull */ ref CFStreamClientContext clientContext);
+
 
-		public static CFRunLoopSource ExecuteProxyAutoConfigurationURL (NSUrl proxyAutoConfigurationURL, NSUrl targetURL, CFProxyAutoConfigurationResultCallback resultCallback, CFStreamClientContext clientContext)
+		public static CFProxy[] ExecuteProxyAutoConfigurationUrl (Uri proxyAutoConfigurationUrl, Uri targetUrl)
 		{
-			if (proxyAutoConfigurationURL == null)
-				throw new ArgumentNullException ("proxyAutoConfigurationURL");
+			if (proxyAutoConfigurationUrl == null)
+				throw new ArgumentNullException (nameof (proxyAutoConfigurationUrl));
 
-			if (targetURL == null)
-				throw new ArgumentNullException ("targetURL");
+			if (targetUrl == null)
+				throw new ArgumentNullException (nameof (targetUrl));
+
+			using (var pacUrl = new NSUrl (proxyAutoConfigurationUrl.AbsoluteUri)) // toll free bridge to CFUrl
+			using (var url = new NSUrl (targetUrl.AbsoluteUri)) {
+				if (pacUrl == null)
+					return null;
 
-			if (resultCallback == null)
-				throw new ArgumentNullException ("resultCallback");
+				if (url == null)
+					return null;
 
-			if (clientContext == null)
-				throw new ArgumentNullException ("clientContext");
+				CFProxy[] proxies = null;
+				var runLoop = CFRunLoop.Current;
 
-			IntPtr source = CFNetworkExecuteProxyAutoConfigurationURL (proxyAutoConfigurationURL.Handle, targetURL.Handle, resultCallback, clientContext);
-			return (source == IntPtr.Zero) ? null : new CFRunLoopSource (source);
+				CFProxyAutoConfigurationResultCallback cb = delegate (IntPtr client, IntPtr proxyList, IntPtr error) {
+					if (proxyList != IntPtr.Zero) {
+						var array = new CFArray (proxyList, false);
```

Comment:
`using (var array = new CFArray (proxyList, false)) {` avoids the manual Dispose later.